cmd/openfga/openfga_test.go

```diff
@@ -0,0 +1,919 @@
+	t.Helper()
+
+	dockerClient, err := client.NewClientWithOpts(client.FromEnv)
+	require.NoError(t, err)
+
+	cmd := []string{"run"}
+	cmd = append(cmd, args...)
+
+	containerCfg := container.Config{
+		Env: []string{},
+		ExposedPorts: nat.PortSet{
+			nat.Port("8080/tcp"): {},
+			nat.Port("8081/tcp"): {},
+			nat.Port("3000/tcp"): {},
+		},
+		Image: "openfga/openfga",
+		Cmd:   cmd,
+	}
+
+	hostCfg := container.HostConfig{
+		AutoRemove:      true,
+		PublishAllPorts: true,
+		PortBindings: nat.PortMap{
+			"8080/tcp": []nat.PortBinding{},
+			"8081/tcp": []nat.PortBinding{},
+			"3000/tcp": []nat.PortBinding{},
+		},
+	}
+
+	ulid, err := id.NewString()
+	require.NoError(t, err)
+
+	name := fmt.Sprintf("openfga-%s", ulid)
+
+	cont, err := dockerClient.ContainerCreate(context.Background(), &containerCfg, &hostCfg, nil, nil, name)
+	require.NoError(t, err, "failed to create openfga docker container")
+
+	stopContainer := func() {
+
+		timeout := 5 * time.Second
+
+		err := dockerClient.ContainerStop(context.Background(), cont.ID, &timeout)
+		if err != nil && !client.IsErrNotFound(err) {
+			t.Fatalf("failed to stop openfga container: %v", err)
+		}
+	}
+
+	err = dockerClient.ContainerStart(context.Background(), cont.ID, types.ContainerStartOptions{})
+	require.NoError(t, err)
+
+	t.Cleanup(stopContainer)
+
+	// spin up a goroutine to survive any test panics or terminations to expire/stop the running container
+	go func() {
+		time.Sleep(2 * time.Minute)
+
+		// swallow the error because by this point we've terminated
+		_ = dockerClient.ContainerStop(context.Background(), cont.ID, nil)
+	}()
+
+	containerJSON, err := dockerClient.ContainerInspect(context.Background(), cont.ID)
+	require.NoError(t, err)
+
+	ports := containerJSON.NetworkSettings.Ports
+
+	m, ok := ports["8080/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get HTTP host port mapping from openfga container")
+	}
+	httpPort := m[0].HostPort
+
+	m, ok = ports["8081/tcp"]
+	if !ok || len(m) == 0 {
+		t.Fatalf("failed to get grpc host port mapping from openfga container")
+	}
+	grpcPort := m[0].HostPort
+
+	timeoutCtx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+	defer cancel()
+
+	creds := insecure.NewCredentials()
+	// if transportCredentials != nil {
+	// 	creds = transportCredentials
+	// }
+
+	dialOpts := []grpc.DialOption{
+		grpc.WithBlock(),
+		grpc.WithTransportCredentials(creds),
+		grpc.WithConnectParams(grpc.ConnectParams{Backoff: grpcbackoff.DefaultConfig}),
+	}
+
+	conn, err := grpc.DialContext(
+		timeoutCtx,
+		fmt.Sprintf("localhost:%s", grpcPort),
+		dialOpts...,
+	)
+	require.NoError(t, err)
+	defer conn.Close()
+
+	client := healthv1pb.NewHealthClient(conn)
+
+	backoffPolicy := backoff.NewExponentialBackOff()
+	backoffPolicy.MaxElapsedTime = 30 * time.Second
+
+	err = backoff.Retry(func() error {
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
```

Comment:
Do we need this? Or should we just set `MaxElapsedTime` to 2 seconds above?

Reply:
I don't want the `Check` to hang until `MaxElapsedTime` has been reached. I want to limit how long the Check can take and then retry the check again in a subsequent retry. If we didn't set the timeout ctx then the single Check could hang until `MaxElapsedTime` elapsed.